src/main/java/com/google/devtools/build/lib/exec/CompactSpawnLogContext.java

```diff
@@ -471,83 +523,100 @@ private int logDirectory(
                     ExecLogEntry.Directory.newBuilder()
                         .setPath(input.getExecPathString())
                         .addAllFiles(
-                            expandDirectory(root, /* pathPrefix= */ null, inputMetadataProvider))));
+                            expandDirectory(root, /* pathPrefix= */ inputMetadataProvider))));
   }
 
   /**
-   * Logs a runfiles directory.
+   * Logs a runfiles directory by storing the information in its {@link RunfilesTree}.
    *
-   * <p>We can't use {@link #logDirectory} because the runfiles symlink tree might not have been
-   * materialized on disk. We must follow the mappings to the actual location of the artifacts.
+   * <p>Since runfiles trees can be very large and, for tests, are only used by a single spawn, we
+   * store them in the log as a special entry that references the nested set of artifacts instead of
+   * as a flat directory.
    *
-   * @param root the path to the runfiles directory
-   * @param mapping a map from runfiles-relative path to the underlying artifact, or null for an
-   *     empty file
-   * @return the entry ID of the {@link ExecLogEntry.Directory} describing the directory.
+   * @param shared whether this runfiles tree is likely to be contained in more than one Spawn's
+   *     inputs
+   * @return the entry ID of the {@link ExecLogEntry.RunfilesTree} describing the directory.
    */
-  private int logRunfilesDirectory(
-      PathFragment root,
-      Map<PathFragment, Artifact> mapping,
+  private int logRunfilesTree(
+      RunfilesTree runfilesTree,
       InputMetadataProvider inputMetadataProvider,
-      FileSystem fileSystem)
+      FileSystem fileSystem,
+      boolean shared)
       throws IOException, InterruptedException {
     return logEntry(
-        root.getPathString(),
+        shared ? runfilesTree.getExecPath().getPathString() : null,
```

Comment:
I can't attach this review comment there, but can you amend the comment in line 148 (describing the purpose of entryMap) to also mention "runfiles tree"?